File: SampleCode/VisaAPICalls/Python/CybersourceSample.py
import requests
import calendar
import time
import datetime
from hashlib import sha256

def generate_timestamp():
    d = datetime.datetime.utcnow()
    timestamp = calendar.timegm(d.timetuple())
    return str(timestamp)

def generate_x_pay_token(shared_secret, resource_path, query_params, body):
    timestamp = generate_timestamp()
    pre_hash_string = shared_secret + timestamp + resource_path + query_params + body
    hash_string = sha256(pre_hash_string.rstrip()).hexdigest()
    return 'x:' + timestamp + ':' + hash_string

def authorize_credit_card(shared_secret, api_key):
    base_uri = 'cybersource/'
    resource_path = 'payments/v1/authorizations'
    query_params = 'apikey=' + api_key
    body = '''{
        "amount": "0",
        "currency": "USD",
        "payment": {
        "cardNumber": "4111111111111111",
        "cardExpirationMonth": "10",
        "cardExpirationYear": "2016"
        }
    }'''
    x_pay_token = generate_x_pay_token(shared_secret, resource_path, query_params, body)	
    headers = {'content-type': 'application/json', 'x-pay-token': x_pay_token }
    r = requests.post('https://sandbox.api.visa.com/' + base_uri + resource_path + '?' + query_params, 
        headers = headers,
        data = body)
    print r.status_code
    print r.content

def main():
    api_key = 'put your api key here'
    shared_secret = 'put your shared secret here'	
    authorize_credit_card(shared_secret, api_key)

if __name__ == '__main__':
    main()

File: SampleCode/VisaAPICalls/Node/CybersourceSample.js
var request = require('request');

var apiKey = 'put your api key here';
var sharedSecret = 'put your shared secret here';
var baseUri = 'cybersource/';
var resourcePath = 'payments/v1/authorizations';
var queryParams = 'apikey=' + apiKey;

var postBody = JSON.stringify({
    "amount": "0",
    "currency": "USD",
    "payment": {
      "cardNumber": "4111111111111111",
      "cardExpirationMonth": "10",
      "cardExpirationYear": "2016"
    }
});

var timestamp = Math.floor(Date.now() / 1000);
var preHashString = sharedSecret + timestamp + resourcePath + queryParams + postBody;
var crypto = require('crypto');
var hashString = crypto.createHash('sha256').update(preHashString).digest('hex');
var xPayToken = 'x:' + timestamp + ':' + hashString;


var req = request.defaults();
req.post({
    uri : 'https://sandbox.api.visa.com/' + baseUri + resourcePath + '?' + queryParams,
    headers: {
      'content-type' : 'application/json',
      'x-pay-token' : xPayToken
    },
    body: postBody
  }, function(error, response, body) {
    if (!error) {
      console.log("Response Code: " + response.statusCode);
      console.log("Headers:");
      for(var item in response.headers) {
        console.log(item + ": " + response.headers[item]);
      }
      console.log("Body: "+ body);
    } else {
      console.log("Got error: " + error.message);
    }
  }
);  